x/superfluid/keeper/intermediary_account.go

```diff
@@ -107,10 +110,76 @@ func (k Keeper) SetIntermediaryAccount(ctx sdk.Context, acc types.SuperfluidInte
 	prefixStore.Set(acc.GetAccAddress(), bz)
 }
 
-func (k Keeper) DeleteIntermediaryAccount(ctx sdk.Context, address sdk.AccAddress) {
+func (k Keeper) DeleteAllEmptyIntermediaryAccounts(ctx sdk.Context) {
+	itermedairyAccounts := k.GetAllIntermediaryAccounts(ctx)
+	for _, intermediaryAccount := range itermedairyAccounts {
+		k.DeleteIntermediaryAccountIfNoDelegation(ctx, intermediaryAccount)
+	}
+}
+
+// DeleteIntermediaryAccount deletes given intermediary account from store
+// Note that intermediary account is highly related to staking and delgation, and this
+// method should be used with caution.
+func (k Keeper) DeleteIntermediaryAccountIfNoDelegation(ctx sdk.Context, intermedairyAcc types.SuperfluidIntermediaryAccount) {
+	// check if any delegations or intermediary connections exist
+	if k.IntermediaryAccountDelegationsExists(ctx, intermedairyAcc) {
+		return
+	}
+
 	store := ctx.KVStore(k.storeKey)
+
+	// store for intermediary account
 	prefixStore := prefix.NewStore(store, types.KeyPrefixIntermediaryAccount)
-	prefixStore.Delete(address)
+	prefixStore.Delete(intermedairyAcc.GetAccAddress())
+}
+
+// IntermediaryAccountDelegationsExists returns true if the gien intermediary account has any delegations.
+// We check this by
+// - using staking keeper to check if actual delegations exist,
+// - checking if there are any intermediary account connection remaining to the intermediaryAcc
+// - if there's no synthetic lock with the intermediaryAccount.Denom + intermediaryAccount.ValAddr combination
```

Comment:
Another nit: I don't think that the method spec should be concerned about such implementation details. The API users only need to know what the method does / returns given the inputs.

The caller should not need to know such implementation details, especially when there are comments directly in code that say the same thing.

Please consider removing method spec implementation details while keeping the comments in code.

I'm not feeling strongly about this so feel free to resolve it if you disagree

Reply:
agreed with you that method spec does not need these information, but still thought having organized list of checks would be a useful comment to have. Moved these comments inside methods!